### ISSUE TITLE
throw=False not working with gradient

### ISSUE THREAD

**Issue by matillda123** [google/lineax#243]
Hey, 

im not sure this is intended or a bug from lineax. When i try to get the gradient of an iterative solver with throw=False, the unsuccessful solve is still raised as an error:

```
import lineax
import equinox
import jax.numpy as jnp
import numpy as np

def test(A, b, solver):
    sol = lineax.linear_solve(A, b, solver, throw=False)
    return sol.value


A = lineax.MatrixLinearOperator(jnp.asarray(np.random.uniform(size=(10,10))), tags=lineax.positive_semidefinite_tag)
b = jnp.asarray(np.random.uniform(size=10))

rtol, atol = 1e-12, 1e-12
solvers_iterative = [lineax.CG, lineax.BiCGStab, lineax.GMRES]
solver = solvers_iterative[2](rtol=rtol, atol=atol, max_steps=10)

grad = equinox.filter_grad(lambda A: jnp.sum(test(A, b, solver)))(A)
```

> equinox._errors._EquinoxRuntimeError: The maximum number of solver steps was reached. Try increasing `max_steps`.

I am on: 
JAX: 0.10.2
jaxlib: 0.10.2
Equinox: 0.13.8
Lineax: 0.1.1

**Comment by jpbrodrick89** [comment]
Hi @matillda123 , this is because we set `throw=True` unconditionally on the reverse pass, I'm not very familiar with the reasoning but its apparently because "we have nowhere to pipe the error to". I suggest either loosening your tolerance turning off errors globally with`EQX_ON_ERROR=none` for now, but I'll try think a bit more about this.

Note that if your solver hasn't converged your gradient is also likely to be wrong (the jvp rule assumes Ax=b exactly) so you need to be careful .

If instead you find that you are converging on the forward pass but not on the reverse pass please let me know as that is a bit more interesting and may be due to things such as poor initial guesses on the reverse pass, but just guessing here.

**Comment by patrick-kidger** [comment]
@jpbrodrick89, just to elucidate a little more what is meant by my quote there. To clarify, the point of `throw=False` is not to silently ignore the error – but rather, to allow the user to programatically inspect a failure via `lx.linear_solve(...).result` instead. I argue that silently ignoring errors would be a Very Bad Thing to allow.

But if an error occurs on the backward pass, and we have `throw=False`, then... the user has nowhere they can hook into to check `.result`. The whole thing occurs in codegen'd code.

_Inside baseball: the 'ideal' solution would definitely be to handle errors via some kind of effectful system, so that a user can extract errors out of an arbitrary program, but at minimum that means writing a custom jaxpr interpreter. Whilst various folks have tried doing this for error handling, I don't know of any versions that are genuinely robust (even the built-in JAX one failed on stuff like doubly-nested while loops for a while; I don't know if that's still true). The beautify of `equinox.error_if` basically ended up being that it *didn't* require this custom interprete._
